### PR TITLE
Implement Hash#transform_values and #transform_values! from Ruby 2.4

### PR DESCRIPTION
**Pull request by havenwood**

### CODE REVIEW

File: src/main/ruby/core/hash.rb [hunk 1]
```diff
@@ -470,4 +470,29 @@ def to_proc
   def merge_fallback(other, &block)
     merge(Truffle::Type.coerce_to other, Hash, :to_hash, &block)
   end
+
+  def transform_values
+    if block_given?
```
Comment (eregon) [inline src/main/ruby/core/hash.rb]: Could you use
```ruby
return to_enum(:transform_values) { size } unless block_given?
```
which seems the dominant pattern in the core?

File: src/main/ruby/core/hash.rb [hunk 2]
```diff
@@ -470,4 +470,29 @@ def to_proc
...
+      end
+      h
+    else
+      to_enum(:transform_values) { size }
+    end
+  end
+
+  def transform_values!
+    if block_given?
+      raise RuntimeError, "can't modify frozen Hash" if frozen?
```
Comment (eregon) [inline src/main/ruby/core/hash.rb]: The `self[key]` should already perform a frozen check, isn't it enough?

Comment (havenwood) [reply src/main/ruby/core/hash.rb]: It's enough unless the Hash is empty. It could be `if empty? && frozen?`.

Comment (eregon) [reply src/main/ruby/core/hash.rb]: I see. There is `Truffle.check_frozen` for these cases of eager frozen checks.
Could you use it?
It avoids hard-coding the frozen error message and is a bit lighter than `if frozen?`.

Comment (havenwood) [reply src/main/ruby/core/hash.rb]: Ah, nice. Done.

**Comment by eregon** [comment]
Looks great, thank you for the new methods!

**Pull request merged**